node/block_exchange.go

```diff
@@ -0,0 +1,233 @@
+	blockStore   *goheaderstore.Store[*types.Block]
+	syncerStatus *SyncerStatus
+
+	logger log.Logger
+	ctx    context.Context
+}
+
+func NewBlockExchangeService(ctx context.Context, store ds.TxnDatastore, conf config.NodeConfig, genesis *tmtypes.GenesisDoc, p2p *p2p.Client, logger log.Logger) (*BlockExchangeService, error) {
+	if genesis == nil {
+		return nil, errors.New("genesis doc cannot be nil")
+	}
+	if p2p == nil {
+		return nil, errors.New("p2p client cannot be nil")
+	}
+	// store is TxnDatastore, but we require Batching, hence the type assertion
+	// note, the badger datastore impl that is used in the background implements both
+	storeBatch, ok := store.(ds.Batching)
+	if !ok {
+		return nil, errors.New("failed to access the datastore")
+	}
+	ss, err := goheaderstore.NewStore[*types.Block](storeBatch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the block store: %w", err)
+	}
+
+	return &BlockExchangeService{
+		conf:         conf,
+		genesis:      genesis,
+		p2p:          p2p,
+		ctx:          ctx,
+		blockStore:   ss,
+		logger:       logger,
+		syncerStatus: new(SyncerStatus),
+	}, nil
+}
+
+func (bExService *BlockExchangeService) initBlockStoreAndStartSyncer(ctx context.Context, initial *types.Block) error {
+	if initial == nil {
+		return fmt.Errorf("failed to initialize the blockstore and start syncer")
+	}
+	if err := bExService.blockStore.Init(ctx, initial); err != nil {
+		return err
+	}
+	if err := bExService.syncer.Start(bExService.ctx); err != nil {
+		return err
+	}
+	bExService.syncerStatus.setStarted()
+	return nil
+}
+
+// Initialize block store if needed and broadcasts provided block.
+// Note: Only returns an error in case block store can't be initialized. Logs error if there's one while broadcasting.
+func (bExService *BlockExchangeService) writeToBlockStoreAndBroadcast(ctx context.Context, block *types.Block) error {
+	// For genesis block initialize the store and start the syncer
+	if block.Height() == bExService.genesis.InitialHeight {
+		if err := bExService.blockStore.Init(ctx, block); err != nil {
+			return fmt.Errorf("failed to initialize block store")
+		}
+
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			return fmt.Errorf("failed to start syncer after initializing block store")
+		}
+	}
+
+	// Broadcast for subscribers
+	if err := bExService.sub.Broadcast(ctx, block); err != nil {
+		bExService.logger.Error("failed to broadcast block", "error", err)
+	}
+	return nil
+}
+
+func (bExService *BlockExchangeService) isInitialized() bool {
+	return bExService.blockStore.Height() > 0
+}
+
+// OnStart is a part of Service interface.
+func (bExService *BlockExchangeService) Start() error {
+	// have to do the initializations here to utilize the p2p node which is created on start
+	ps := bExService.p2p.PubSub()
+	bExService.sub = goheaderp2p.NewSubscriber[*types.Block](ps, pubsub.DefaultMsgIdFn, bExService.genesis.ChainID+"-block")
```

Comment:
can use variables for: `bExService.genesis.ChainID+"-block"` and `network+"-block"`